homework/ba-nguyen/lesson-10/js-5.md

```diff
@@ -0,0 +1,274 @@
+    activeTab = event.target;
+    activeTab.classList.add("selected");
+    let tabContent = document.querySelectorAll("div");
+    // Hides the content of the previously selected tab.
+    currentBody.classList.toggle("hidden");
+
+    // Manages display of the contents of the multiple tabbed sections.
+    switch (activeTab.textContent) {
+      case "one":
+        currentBody = tabContent[0];
+        break;
+      case "two":
+        currentBody = tabContent[1];
+        break;
+      case "three":
+        currentBody = tabContent[2];
+        break;
+    }
+    // Displays the content of the currently selected tab.
+    currentBody.classList.toggle("hidden");
+  }
+
+  function asTabs(node) {
+    let tabLength = node.children.length;
+
+    for (let i = 0; i < tabLength; i++) {
+      let tabButton = document.createElement("button");
+
+      tabButton.textContent = document
+        .querySelectorAll("div")
+        [i].getAttribute("data-tabname");
+      // By default, hide the body of the tabs.
+      node.querySelectorAll("div")[i].classList.add("hidden");
+      tabButton.addEventListener("click", switchTab);
+      node.insertBefore(tabButton, document.querySelector("div"));
+    }
+    // Sets the initially selected tab & display its associated content.
+    activeTab = document.querySelector("button");
+    activeTab.classList.add("selected");
+    currentBody = document.querySelector("div");
+    currentBody.classList.toggle("hidden");
+  }
+
+  asTabs(document.querySelector("tab-panel"));
+</script>
+```
+
+**Question 4:**
+
+```css
+input[type="range"] {
+  accent-color: #f44336;
+}
+```
+
+**Question 5: What's the output?**
+
+```js
+const myPromise = Promise.resolve(Promise.resolve("Promise"));
+
+function funcOne() {
+  setTimeout(() => console.log("Timeout 1!"), 0);
+  myPromise.then((res) => res).then((res) => console.log(`${res} 1!`));
+  console.log("Last line 1!");
+}
+
+async function funcTwo() {
+  const res = await myPromise;
+  console.log(`${res} 2!`);
+  setTimeout(() => console.log("Timeout 2!"), 0);
+  console.log("Last line 2!");
+}
+
+funcOne();
+funcTwo();
+```
+
+- A: Promise 1! Last line 1! Promise 2! Last line 2! Timeout 1! Timeout 2!
+- B: Last line 1! Timeout 1! Promise 1! Last line 2! Promise2! Timeout 2!
+- `C: Last line 1! Promise 2! Last line 2! Promise 1! Timeout 1! Timeout 2!` => **Answer**
+- D: Timeout 1! Promise 1! Last line 1! Promise 2! Timeout 2! Last line 2!
+
+**Question 6: How can we invoke sum in sum.js from index.js?**
+
+```js
+// sum.js
+export default function sum(x) {
+  return x + x;
+}
+
+// index.js
+import * as sum from "./sum";
+```
+
+- A: sum(4)
+- B: sum.sum(4)
+- `C: sum.default(4)` => **Answer**
+- D: Default aren't imported with \*, only named exports
+
+**Question 7: What's the output?**
+
+```js
+const handler = {
+  set: () => console.log("Added a new property!"),
+  get: () => console.log("Accessed a property!"),
+};
+
+const person = new Proxy({}, handler);
```

Comment:
Thaycacac: can you tell me this Proxy in reality applied for?

Reply:
some example for you [https://blog.logrocket.com/practical-use-cases-for-javascript-es6-proxies/](https://blog.logrocket.com/practical-use-cases-for-javascript-es6-proxies/)